db/view/view.cc

```diff
@@ -1863,12 +1863,23 @@ future<> view_update_generator::mutate_MV(
     // TODO: Maybe allow users to set use_legacy_self_pairing explicitly
     // on a view, like we have the synchronous_updates_flag.
     bool use_legacy_self_pairing = !ks.uses_tablets();
-    auto base_ermp = base->table().get_effective_replication_map();
+    std::unordered_map<table_id, locator::effective_replication_map_ptr> erms;
+    auto get_erm = [&] (table_id id) {
+        auto it = erms.find(id);
+        if (it == erms.end()) {
+            it = erms.emplace(id, _db.find_column_family(id).get_effective_replication_map()).first;
+        }
+        return it->second;
+    };
+    auto base_ermp = get_erm(base->id());
+    for (const auto& mut : view_updates) {
+        (void)get_erm(mut.s->id());
+    }
     static constexpr size_t max_concurrent_updates = 128;
     co_await utils::get_local_injector().inject("delay_before_get_view_natural_endpoint", 8000ms);
     co_await max_concurrent_for_each(view_updates, max_concurrent_updates, [&] (frozen_mutation_and_schema mut) mutable -> future<> {
         auto view_token = dht::get_token(*mut.s, mut.fm.key());
-        auto view_ermp = mut.s->table().get_effective_replication_map();
+        auto view_ermp = erms.at(mut.s->id());
```

Comment:
I'd expect the topology object to be the same (with regards to the set of nodes in the cluster), while the respective tablet maps for the base and view may be updated separately as you said.

Reply:
Regarding how beneficial it is, I hit a case during local testing where IIRC a host_id was present as a view replica but it was no longer in the base topology.

How can pairing work in this case?

Reply:
> I'd expect the topology object to be the same (with regards to the inventory of nodes in the cluster), while the respective tablet maps for the base and view may be updated separately as you said.

From what I see, `table::update_effective_replication_map` overwrites table's `_erm` and both both topology and tablet map are derived from the `token_metadata` in the `erm`, so we can't assure that the topologies will be the same without assuring the `erm`s are updated at the same time.
 
> Regarding how beneficial it is, I hit a case during local testing where IIRC a host_id was present as a view replica but it was no longer in the base topology.
> 
> How can pairing work in this case?

In this case the new replica that replaces this `host_id` should be pending, so we should send an update to it anyway, even though we just paired ourselves with an outdated replica. Is that what you were referring to?

While thinking about this, I think I actually found another way a view update may get lost during a view tablet migration with rack-aware pairing.
If the migration is set as completed on all replicas aside from one and that replica changes its paired view replica to the some other view replica (not to the target of the migration), we won't send the update to the view replica that gets paired with this "slower" base replica after the migration - it wasn't pending and the originally paired base replica sent its update to another view replica.

I don't see a simple way of fixing this, so this will probably be just another "small" hit to our view consistency. Especially if we use the best-match pairing algorithm - in that case the number of missed view replicas should be at most 1

Reply:
@wmitros so any objection to the above hunk, or can we resolve the conversation?

Reply:
This hunk shouldn't break anything, but I don't think it helps with anything either.

If we were updating base and view effective replication maps together atomically, this patch would prevent us from performing the ERM update between getting base and view ERMs, and that would achieve what we want - having the same version of topology for base and view.

However, we don't do that atomically so the topologies of base and view can still be of different versions, and this code will suggest that we do have the same base and view topology versions, which might result in a new problem like https://github.com/scylladb/scylladb/issues/18709 after some future change. So maybe it would be best if we didn't include this hunk in the series after all.

Reply:
I'd also prefer if we used the same base and view topology during view updates but for that we need the atomic ERM updates and I don't think we have an open issue about it. 

I just looked at the code and perhaps it wouldn't be a difficult change (though I won't suggest to include it in the series):
Currently the ERM update is not atomic because we can yield between updating ERMs of different tables in:
https://github.com/scylladb/scylladb/blob/841481c202ac7674c9524f3fbb80ecf8bf609b27/service/storage_service.cc#L3149-L3151
But in the `update_effective_replication_map` we actually replace the `_erm` synchronously, the whole function is asynchronous because `tablet_storage_group_manager::update_effective_replication_map` is asynchronous:
https://github.com/scylladb/scylladb/blob/841481c202ac7674c9524f3fbb80ecf8bf609b27/replica/table.cc#L2484-L2494
So we could split the loop in `storage_service::replicate_to_all_cores` into two: one that just overwrites the `_erm` synchronously and another that performs the rest of the work.

Reply:
On top of that, I don't like that we may yield after updating _erm and before completing `_sg_manager->update_effective_replication_map` for tablets.
Maybe the right thing to do is to make `table::get_effective_replication_map` async so it can block while the update is ongoing.

Reply:
Dropped "view: mutate_MV: lookup base and view erms synchronously" in v9 (https://github.com/scylladb/scylladb/commit/86cc9da4a15a28a81dd352f7efae9dada751d432)